pageserver/src/storage_sync.rs

```diff
@@ -1573,335 +988,184 @@ mod test_utils {
-            let local_files =
-                HashMap::from([(PathBuf::from("first_file"), LayerFileMetadata::new(123))]);
-            let mut remote_entry = RemoteTimeline::new(local_metadata.clone());
-            remote_entry
-                .add_timeline_layers([(PathBuf::from("first_file"), LayerFileMetadata::new(123))]);
-
-            let (status, sync_needed) = compare_local_and_remote_timeline(
-                &mut new_sync_tasks,
-                sync_id,
-                local_metadata.clone(),
-                local_files,
-                &remote_entry,
-            );
+    // Test scheduling
+    #[test]
+    fn upload_scheduling() -> anyhow::Result<()> {
+        let harness = TenantHarness::create("upload_scheduling")?;
+        let timeline_path = harness.timeline_path(&TIMELINE_ID);
+        std::fs::create_dir_all(&timeline_path)?;
 
-            assert_eq!(
-                status,
-                LocalTimelineInitStatus::LocallyComplete(local_metadata)
-            );
-            assert!(!sync_needed);
+        let remote_fs_dir = harness.conf.workdir.join("remote_fs");
+        std::fs::create_dir_all(remote_fs_dir)?;
+        let remote_fs_dir = std::fs::canonicalize(harness.conf.workdir.join("remote_fs"))?;
 
-            assert!(new_sync_tasks.is_empty(), "{:?}", new_sync_tasks);
-        }
-
-        #[test]
-        fn needs_download() {
-            let mut new_sync_tasks = VecDeque::default();
-            let sync_id = TenantTimelineId::generate();
-            let local_metadata = dummy_metadata(0x02.into());
-            let local_files = HashMap::default();
-            let mut remote_entry = RemoteTimeline::new(local_metadata.clone());
-            remote_entry
-                .add_timeline_layers([(PathBuf::from("first_file"), LayerFileMetadata::new(123))]);
-
-            let (status, sync_needed) = compare_local_and_remote_timeline(
-                &mut new_sync_tasks,
-                sync_id,
-                local_metadata,
-                local_files.clone(),
-                &remote_entry,
-            );
-
-            assert_eq!(status, LocalTimelineInitStatus::NeedsSync);
-            assert!(sync_needed);
+        let storage_config = RemoteStorageConfig {
+            max_concurrent_syncs: std::num::NonZeroUsize::new(
+                remote_storage::DEFAULT_REMOTE_STORAGE_MAX_CONCURRENT_SYNCS,
+            )
+            .unwrap(),
+            max_sync_errors: std::num::NonZeroU32::new(
+                remote_storage::DEFAULT_REMOTE_STORAGE_MAX_SYNC_ERRORS,
+            )
+            .unwrap(),
+            storage: RemoteStorageKind::LocalFs(remote_fs_dir.clone()),
+        };
 
-            let new_sync_tasks = new_sync_tasks.into_iter().collect::<Vec<_>>();
+        // Use a current-thread runtime in the test
+        let runtime = Box::leak(Box::new(
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()?,
+        ));
```

Comment:
```suggestion
        let runtime = 
            tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()?;
```

I don't see why this should be leaked. `current_thread` will only advance futures during the `block_on` regardless. Or just `#[tokio::test]`.

Reply:
RemoteTimelineClient needs a static reference to the runtime: `runtime: &'static Runtime`. We use Box::leak to get a static reference.

I'm sure you could arrange it differently, e.g. by changing `RemoteTimelineClient` to `RemoteTimelineClient<'T>`, but I'd prefer to have some ugliness like Box::leak in unit test code, and keep the rest of the code simpler.